### DISCUSSION TITLE
Estimate parameters for several within-subject conditions

### DISCUSSION THREAD

**Discussion by SadeghGhasemian** [mwshinn/PyDDM#75] [Ask for help]
When we estimate parameters for several within-subject conditions (e.g., coherence levels in Roitman data) the output is an average (across condition) for all parameters. Is it possible to retrieve parameters for separate conditions in PyDDM. 

Bests,

Sadegh

**Comment by mwshinn** [comment]
I'm not quite sure what you mean here.  Isn't it possible to use the input parameters to compute the output?  (E.g. in the Roitman example, if driftcoh = .5, then for coh=.1 is is .5*.1=.05, for coh = .5 it is .5*.5, etc.)

**Comment by SadeghGhasemian** [comment]
*Thank you for your quick reply. *

*This is code that I'm using to fit the model to the data from a Temporal
Bisection task. I used a range of 'duration' (0.2, 0.3, 0.4, 0.5, 0.6, 0.7,
0.8s), similar to the 'coherence' in Roitman study. In Each trial a test
duration is presented and participants choice left or right bottom/key
based on the proximity of that duration to anchor durations (0.2; short
anchor and 0.8; long anchor). Here we should allow the 'IC' and 'Drift'
rate to vary with 'duration' * *because both 'IC' and 'Drift' substantially
change with duration**.  The following codes were used for parameter
estimation: *

import pandas
import numpy as np
import pyddm.plot as PL
import matplotlib.pyplot as plt
from pyddm.plot import model_gui
import pyddm as ddm
from pyddm import Model, Fittable, Sample, InitialCondition,
NoiseConstant,BoundConstant, OverlayChain, OverlayNonDecision,
OverlayPoissonMixture, LossRobustLikelihood, LossRobustBIC
from pyddm.functions import fit_adjust_model, display_model
with open("TCSubb2.csv", "r") as f:
    df_rt = pandas.read_csv(f)
#df_rt = df_rt[df_rt["subject"] == 1] # 800ms
df_rt = df_rt[df_rt["rt"] > 0.1] # Remove trials less than 100ms
df_rt = df_rt[df_rt["rt"] < 1.6] # Remove trials greater than 1650ms
B_sample = Sample.from_pandas_dataframe(df_rt, rt_column_name="rt",
correct_column_name="correct")
class ICPointSideBias(InitialCondition):
    name = "A starting point with a left or right bias."
    required_parameters = ["x0"]
    required_conditions = ["duration"]
    def get_IC(self, x, dx, conditions):
        start = np.round(self.x0/dx)
        # Positive bias for left choices, negative for right choices
        if not conditions['duration']:
            start = -start
        shift_i = int(start + (len(x)-1)/2)
        assert shift_i >= 0 and shift_i < len(x), "Invalid initial
conditions"
        pdf = np.zeros(len(x))
        pdf[shift_i] = 1. # Initial condition at x=self.x0.
        return pdf
class DriftDurationLeak(ddm.models.Drift):
    name = "Leaky drift depends linearly on duration"
    required_parameters = ["driftduration", "leak"] # <-- Parameters we
want to include in the model
    required_conditions = ["duration"] # <-- Task parameters
("conditions"). Should be the same name as in the sample.
    # We must always define the get_drift function, which is used to
compute the instantaneous value of drift.
    def get_drift(self, x, conditions, **kwargs):
        return self.driftduration + self.leak * x
B_model = Model(name=' Bisection data, drift varies with duration', drift =
DriftDurationLeak(driftduration=Fittable(minval=-50, maxval=50),
                                            leak=Fittable(minval=0, maxval=
10)), noise=NoiseConstant(noise= 1),
                                            bound=BoundConstant(B=1.5),

IC=ICPointSideBias(x0=Fittable(minval=0, maxval=1.5)),

overlay=OverlayChain(overlays=[OverlayNonDecision(nondectime=Fittable(minval=0,
maxval=.5)),

OverlayPoissonMixture(pmixturecoef=.02,

rate=1)]),

dx=.01, dt=.01, T_dur=2.5)
fit_B_model = fit_adjust_model(sample=B_sample, model=B_model,
lossfunction= LossRobustBIC, verbose=False)
model_gui(fit_B_model, sample=B_sample)
display_model(fit_B_model)
PL.plot_fit_diagnostics(model=fit_B_model, sample=B_sample)
plt.show(fit_B_model)
fit_B_model.parameters()

*And this is the output for one of subjects: *
Info: Params [-1.12853062 4.24977546 0.34260647 0.06619511] gave
658.9800331748418

Model  Bisection data, drift varies with duration information:
Drift component DriftDurationLeak:
    Leaky drift depends linearly on duration
    Fitted parameters:
    - driftduration: -1.128531
    - leak: 4.249775
Noise component NoiseConstant:
    constant
    Fixed parameters:
    - noise: 1.000000
Bound component BoundConstant:
    constant
    Fixed parameters:
    - B: 1.500000
IC component ICPointSideBias:
    A starting point with a left or right bias.
    Fitted parameters:
    - x0: 0.342606
Overlay component OverlayChain:
    Overlay component OverlayNonDecision:
        Add a non-decision by shifting the histogram
        Fitted parameters:
        - nondectime: 0.066195
    Overlay component OverlayPoissonMixture:
        Poisson distribution mixture model (lapse rate)
        Fixed parameters:
        - pmixturecoef: 0.020000
        - rate: 1.000000
Fit information:
    Loss function: BIC
    Loss function value: 658.9800331748418
    Fitting method: differential_evolution
    Solver: auto
    Other properties:
        - nparams: 4
        - samplesize: 553
        - mess: ''

{'drift': {'driftduration': Fitted(-1.128530623358101, minval=-50, maxval=50),
  'leak': Fitted(4.24977546215369, minval=0, maxval=10)},
 'noise': {'noise': 1},
 'bound': {'B': 1.5},
 'IC': {'x0': Fitted(0.3426064675966125, minval=0, maxval=1.5)},
 'overlay': {'nondectime': Fitted(0.06619510904082421, minval=0, maxval=0.5),
  'pmixturecoef': 0.02,
  'rate': 1}}

[image: image.png]
*As you can see in the output, there is only an average value for each of
'IC' and 'Drift'.*
*But, in addition to averaged values I need to have a value for each
duration. Unfortunately, I could not find a way to retrieve duration-wise
values. *
*Is there a way to do that? *

*In the figure 'Correct' and 'Error' represents 'Long' and 'Short' choices,
respectively.*

*Bests,*


*Ghasemian*

**Reply by mwshinn** [reply]
It looks like you are not using your task condition anywhere.  In the Roitman case, you need to use conditions["duration"] within the equation from the get_drift function in whatever way you want the drift to depend on duration.  So it is not returning an average, it is just completely ignoring the duration variable since you don't do anything with it in your get_drift function.

By the way, you also seem to flip your initial condition based on the duration, which I don't think makes sense in your case.  I think what you are trying to do is fit each condition independently?  If so, you can accomplish this with if statements in your get_drift and get_IC functions.
